llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -1230,46 +1233,59 @@ RISCVVLOptimizer::getMinimumVLForUser(MachineOperand &UserOp) {
   // Looking for an immediate or a register VL that isn't X0.
   assert((!VLOp.isReg() || VLOp.getReg() != RISCV::X0) &&
          "Did not expect X0 VL");
+
+  // If we know the demanded VL of UserMI, then we can reduce the VL it
+  // requires.
+  if (DemandedVLs.contains(&UserMI)) {
+    // We can only shrink the demanded VL if the elementwise result doesn't
+    // depend on VL (i.e. not vredsum/viota etc.)
+    // Also conservatively restrict to supported instructions for now.
+    // TODO: Can we remove the isSupportedInstr check?
+    if (!RISCVII::elementsDependOnVL(
+            TII->get(RISCV::getRVVMCOpcode(UserMI.getOpcode())).TSFlags) &&
+        isSupportedInstr(UserMI)) {
```

Comment:
isSupportedInstr is supposed to be the set of instructions that we can reduce VL for. We aren't actually reducing the VL of UserMI here though. I see why there might be significant overlap between what you are querying here and what that function checks. I think we should keep the check for now.

Reply:
Yeah, I had just added out of an abundance of caution. I think `!RISCVII::elementsDependOnVL` is enough for correctness.

Actually now that you mention it, we already check for isSupportedInstr before inserting anything into demandedVLs. So I think this isSupportedInstr check is redundant. Should I remove it?